Add ability to start TensorFlow server from Java API

dmitrievanthony opened the pull request:
[Standalone Client Mode](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/distribute/README.md#standalone-client-mode) allows to easily train model utilizing distributed resources. Only one thing we need to do that is to start TensorFlow server on every cluster node we'd like to participate in training. Because of that it's very important to have an ability to start TensorFlow server in different ways and from any environment. I prepared this request as result of [discussion on Development List](https://groups.google.com/a/tensorflow.org/forum/#!topic/developers/TVtFC94bC8k).

The following example demonstrates how to use TensorFlow server in Java:
```java
ClusterDef clusterDef = ClusterDef.newBuilder()
  .addJob(JobDef.newBuilder()
  .setName("worker")
  .putTasks(0, "localhost:4321")
  .build()
).build();

ServerDef serverDef = ServerDef.newBuilder()
  .setCluster(clusterDef)
  .setJobName("worker")
  .setTaskIndex(0)
  .setProtocol("grpc")
  .build();

Server server = new Server(serverDef.toByteArray());
server.start();
server.join();
```

Please, feel free to comment and suggest changes.

dmitrievanthony committed:
Add ability to start TensorFlow server from Java API.

asimshankar commented on tensorflow/c/c_api.h:
```diff
@@ -1662,6 +1662,29 @@ TF_CAPI_EXPORT extern TF_Buffer* TF_GetAllRegisteredKernels(TF_Status* status);
+// --------------------------------------------------------------------------
+// Server functionality.
+
+// Server.
```
Adding some detail in the comments will be useful, as this is the only form of documentation we have for the C API. Same goes for all other functions added.

asimshankar commented on tensorflow/c/c_api.h:
```diff
@@ -1662,6 +1662,29 @@ TF_CAPI_EXPORT extern TF_Buffer* TF_GetAllRegisteredKernels(TF_Status* status);
+typedef struct TF_Server TF_Server;
+
+// Creates new server.
+TF_CAPI_EXPORT extern TF_Server* TF_NewServer(const void* proto,
```
Here I'd document what "proto" and "proto_len" is (a serialized `ServerDef` proto apparently?).
And also talk about how users should use the returned object.

asimshankar commented on tensorflow/c/c_api.h:
```diff
@@ -1662,6 +1662,29 @@ TF_CAPI_EXPORT extern TF_Buffer* TF_GetAllRegisteredKernels(TF_Status* status);
+                                              TF_Status* status);
+
+// Starts a server.
+TF_CAPI_EXPORT extern void TF_StartServer(TF_Server* server, TF_Status* status);
```
Let's more closely follow the convention of most of the types in this file. So:

- `TF_NewServer`
- `TF_ServerStart`
- `TF_ServerStop`
- `TF_ServerJoin`
- `TF_DeleteServer`

asimshankar commented on tensorflow/c/c_api.h:
```diff
@@ -1662,6 +1662,29 @@ TF_CAPI_EXPORT extern TF_Buffer* TF_GetAllRegisteredKernels(TF_Status* status);
+// Blocks until the server has shut down (currently blocks forever).
+TF_CAPI_EXPORT extern void TF_JoinServer(TF_Server* server, TF_Status* status);
+
+// Destroy a server, frees memory.
```
I'd clarify here whether the server is expected to have been stopped/joined before calling this.

asimshankar commented on tensorflow/c/c_api_internal.h:
```diff
@@ -179,6 +180,12 @@ struct TF_ApiDefMap {
   tensorflow::mutex lock;
 };
 
+struct TF_Server {
```
Ownership of the `server` pointer is probably best expressed using `std::unique_ptr`. So something like:

```c++
struct TF_Server {
  std::unique_ptr<tensorflow::ServerInterface> server;
}
```

asimshankar commented on tensorflow/c/c_api.cc:
```diff
@@ -2803,4 +2803,43 @@ TF_Buffer* TF_GetRegisteredKernelsForOp(const char* name, TF_Status* status) {
+                        TF_Status* status) {
+  tensorflow::ServerDef server_def;
+  if (!server_def.ParseFromArray(proto, static_cast<int>(proto_len))) {
+    status->status = InvalidArgument("Unparseable ServerDef");
```
Let's spell the error out in more detail: "Could not parse provided bytes into a tensorflow.ServerDef protocol buffer."

asimshankar commented on tensorflow/c/c_api.cc:
```diff
@@ -2803,4 +2803,43 @@ TF_Buffer* TF_GetRegisteredKernelsForOp(const char* name, TF_Status* status) {
+    return nullptr;
+  }
+
+  auto out_server = new std::unique_ptr<tensorflow::ServerInterface>();
```
Isn't `out_server` being leaked here since the `std::unique_ptr` is being heap allocated?
See comment in `c_api_internal.h`, I think we can simplify this to:

```c++
std::unique_ptr<TF_Server> ret(new TF_Server);
status->status = tensorflow::NewServer(server_def, &ret->server);
if (!status->status.ok()) return nullptr;
return ret.release();
```

asimshankar commented on tensorflow/c/c_api.cc:
```diff
@@ -2803,4 +2803,43 @@ TF_Buffer* TF_GetRegisteredKernelsForOp(const char* name, TF_Status* status) {
+}
+
+void TF_DeleteServer(TF_Server* server) {
+  if (server != nullptr) {
```
Aren't we leaking the `TF_Server*` pointer? How does that get cleaned up?
See comment in `c_api_internal.h`. If we agree on that, then this can be simplified to:

```c++
void TF_DeleteServer(TF_Server* server) { delete server; }
```

asimshankar commented on tensorflow/java/src/main/java/org/tensorflow/Server.java:
```diff
@@ -0,0 +1,82 @@
+ * {@code tf.Session} target that can participate in distributed training. A
+ * server belongs to a cluster (specified by a {@code tf.train.ClusterSpec}),
+ * and corresponds to a particular task in a named job. The server can
+ * communicate with any other server in the same cluster.
```
Let's add some commentary/sample code on how to use objects of this class (`start()`, `stop()` etc.)

asimshankar commented on tensorflow/java/src/main/native/server_jni.cc:
```diff
@@ -0,0 +1,73 @@
+
+JNIEXPORT jlong JNICALL Java_org_tensorflow_Server_allocate(
+    JNIEnv* env, jclass clazz, jbyteArray server_def) {
+  TF_Status* status = TF_NewStatus();
```
I think we're leaking this, `TF_DeleteStatus` needs to be called at some point.

asimshankar commented on tensorflow/java/src/main/native/server_jni.cc:
```diff
@@ -0,0 +1,73 @@
+JNIEXPORT void JNICALL Java_org_tensorflow_Server_start(JNIEnv* env,
+                                                        jclass clazz,
+                                                        jlong handle) {
+  TF_Status* status = TF_NewStatus();
```
Need corresponding `TF_DeleteStatus` in this function too.

asimshankar commented on tensorflow/java/src/main/native/server_jni.cc:
```diff
@@ -0,0 +1,73 @@
+JNIEXPORT void JNICALL Java_org_tensorflow_Server_stop(JNIEnv* env,
+                                                       jclass clazz,
+                                                       jlong handle) {
+  TF_Status* status = TF_NewStatus();
```
Need corresponding `TF_DeleteStatus` in this function too.

asimshankar commented on tensorflow/java/src/main/native/server_jni.cc:
```diff
@@ -0,0 +1,73 @@
+JNIEXPORT void JNICALL Java_org_tensorflow_Server_join(JNIEnv* env,
+                                                       jclass clazz,
+                                                       jlong handle) {
+  TF_Status* status = TF_NewStatus();
```
Need corresponding `TF_DeleteStatus` in this function too.

asimshankar commented on tensorflow/java/src/main/native/server_jni.cc:
```diff
@@ -0,0 +1,73 @@
+  TF_Status* status = TF_NewStatus();
+  TF_Server* server = reinterpret_cast<TF_Server*>(handle);
+
+  TF_JoinServer(server, status);
```
In all these functions, it is probably prudent to handle the case where `Server.close()` has already been called (and thus `handle` is 0). I think currently this will crash.

You may want to do something like: https://github.com/tensorflow/tensorflow/blob/0f1894de6cfddb3df8a3d6865d97b947af56f995/tensorflow/java/src/main/native/graph_jni.cc#L25

dmitrievanthony replied:
Done.

asimshankar commented on tensorflow/java/src/main/java/org/tensorflow/Server.java:
```diff
@@ -0,0 +1,82 @@
+  }
+
+  @Override
+  public void close() {
```
Having both `stop()` and `close()` seems unnecessary. How about just having `close()`, which invokes `TF_ServerStop`? And `join()` can zero out the `nativeHandle` after returning.

dmitrievanthony replied:
I would prefer to keep these two methods separately. The reason for that is simple, `Server` is `AutoCloseable` and that means that common use case will look like this:

```java
try (Server server = new Server(...)) {
  server.start();
  server.join(); // or server.stop();
}
```

As you can see in this case method `close()` will be called automatically after `join`, so it `join` frees resources itself it will lead to illegal state exception.

asimshankar commented on tensorflow/java/src/main/java/org/tensorflow/Server.java:
```diff
@@ -0,0 +1,82 @@
+  }
+
+  /** Blocks until the server has shut down (currently blocks forever). */
+  public synchronized void join() {
```
Given that `join()` blocks, are you sure we want to mark this method as `synchronized`?
It appears to me, that if one thread were to invoke `join()` first, then no other thread will be able to invoke `stop()` and as a result both threads will remain blocked forever. Am I missing something?

dmitrievanthony replied:
I've checked code and looks like `Server` methods `start`, `stop` and `join` are synchronized on underlying layer (I didn't know that). So, we only need to prohibit parallel calls of `delete` and other functions. Do do that I added rw-lock, so please have a look.

Using this lock, user will be able to call `stop` method from other thread when current thread is locked by `join` method.

dmitrievanthony replied:
By the way, I checked calling `stop` when server is locked by `join` and looks like it doesn't work (see https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/distributed_runtime/rpc/grpc_server_lib.cc#L412)

dmitrievanthony replied:
I investigated a bit more and found out that `close()` just can't work correctly. `Server` destructor calls `stop()` wrapped into `TF_CHECK_OK`, but `stop()` returns `unimplemented` exception, as result `TF_CHECK_OK` crushes the process using sig 6.

Looks like it should be fixed, but in other PR (it's not about Java API at all).

asimshankar commented on tensorflow/java/src/main/native/BUILD:
```diff
@@ -43,6 +43,7 @@ tf_cuda_library(
             "//tensorflow/core:all_kernels",
             "//tensorflow/core:direct_session",
             "//tensorflow/core:ops",
+            "//tensorflow/core/distributed_runtime/rpc:grpc_server_lib",
```
Should we add this dependency to https://github.com/tensorflow/tensorflow/blob/ea718e872df737a6ee69f53a8f9b212c9c8197ec/tensorflow/java/BUILD#L384 instead?

dmitrievanthony replied:
Done.

asimshankar reviewed (changes requested):
Thanks for the PR! A few comments

dmitrievanthony committed:
Server Java API updates after review.

dmitrievanthony commented:
Hi @asimshankar, thank you for review. I fixed all your comments, so please have a look.

dmitrievanthony commented:
Hi @asimshankar. Any updates?

dmitrievanthony commented:
Hi, @asimshankar, @ymodak. I don't clearly understand current state. Do you have some concerns about this PR or it's ready to be merged?

asimshankar commented:
@dmitrievanthony - apologies for the delay, I'll aim to take a look at the updated PR soon

asimshankar reviewed (changes requested):
Thanks for the changes, some additional comments.

Also, I think with your C API change we should be able to use this in Python as well and reduce the custom wrapping (probably there for historical reasons) in https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/training/server_lib.i

If you're up for that in a follow up PR, that would be awesome. Otherwise, I can take a stab at it once this PR is in.

dmitrievanthony commented on tensorflow/c/c_api.h:
```diff
@@ -1662,6 +1662,56 @@ TF_CAPI_EXPORT extern TF_Buffer* TF_GetAllRegisteredKernels(TF_Status* status);
+//  status - Set to OK on success and an appropriate error on failure.
+TF_CAPI_EXPORT extern void TF_ServerStop(TF_Server* server, TF_Status* status);
+
+// Blocks until the server has shut down (currently blocks forever).
```
Done.

dmitrievanthony commented on tensorflow/c/c_api.h:
```diff
@@ -1662,6 +1662,56 @@ TF_CAPI_EXPORT extern TF_Buffer* TF_GetAllRegisteredKernels(TF_Status* status);
+//  status - Set to OK on success and an appropriate error on failure.
+TF_CAPI_EXPORT extern void TF_ServerJoin(TF_Server* server, TF_Status* status);
+
+// Destroy a server, frees memory. Server is expected to be stopped before.
```
Done, please have a look.

dmitrievanthony commented on tensorflow/java/src/main/java/org/tensorflow/Server.java:
```diff
@@ -0,0 +1,134 @@
+ * <p>Using example:
+ * <pre>
+ * {@code
+ * ClusterDef clusterDef = ClusterDef.newBuilder()
```
I think I can't do it because `ClusterDef`, `JobDef` etc. are classes from `org.tensorflow::proto`, not from this package.

asimshankar replied:
This is just in the sample code for Javadoc, so the import statements should be fine?

dmitrievanthony replied:
Got it, done. I just accidentally thought that you mean add imports to provide links to these classes in javadoc.

dmitrievanthony commented on tensorflow/java/src/main/java/org/tensorflow/Server.java:
```diff
@@ -0,0 +1,134 @@
+ * <p>Instances of a {@code Server} are thread-safe.
+ *
+ * <p>Using example:
+ * <pre>
```
Done.

dmitrievanthony commented on tensorflow/java/src/main/java/org/tensorflow/Server.java:
```diff
@@ -0,0 +1,134 @@
+
+  /** Blocks until the server has shut down (currently blocks forever). */
+  public void join() {
+    lock.readLock().lock();
```
Do you think it's better than a simple `RWLock`? I see at least one issue in you example (`notifyAll` should be called instead of `notify`) and two tricky parts (`numJoining` and `wait`/`while`). `RWLock` looks much easier.

asimshankar replied:
Yes should be `notifyAll`. 
What is the danger with the `while` and the `numJoining` counter?

I find the `RWLock` harder to follow, particularly since it is easy to gloss over the subtlety that `join` is a "write" operation.

Isn't this scenario exactly what the object monitor and `wait`/notify mechanism is meant for?

dmitrievanthony replied:
The `numJoining` is danger because it's quite hard to understand if it should be `volatile` in this case or not. Method `wait()` it tricky because it should be wrapped into `while` loop.

Well, anyway, I've been thinking about locks because the problems for me looks like "forbid parallel execution of `close()` and other methods". Anyway, if you think `synchronized/wait/notify` approach is better I'll update the code.

dmitrievanthony replied:
Done.

dmitrievanthony commented on tensorflow/java/src/main/java/org/tensorflow/Server.java:
```diff
@@ -0,0 +1,134 @@
+    }
+  }
+
+  /** Blocks until the server has shut down (currently blocks forever). */
```
Removed because you are right, it's about GRPC server, not API itself.

dmitrievanthony commented on tensorflow/java/src/main/native/server_jni.cc:
```diff
@@ -0,0 +1,105 @@
+JNIEXPORT void JNICALL Java_org_tensorflow_Server_stop(JNIEnv* env,
+                                                       jclass clazz,
+                                                       jlong handle) {
+  if (handle == 0) {
```
Done.

dmitrievanthony commented on tensorflow/java/src/main/native/server_jni.cc:
```diff
@@ -0,0 +1,105 @@
+JNIEXPORT void JNICALL Java_org_tensorflow_Server_delete(JNIEnv* env,
+                                                         jclass clazz,
+                                                         jlong handle) {
+  if (handle == 0) {
```
After moving this code into `requireHandle` function the same logic is applied in all methods. It's aligned with approach used in `TF_Session` as you suggested above.

dmitrievanthony committed:
Update after review.

dmitrievanthony commented:
Hi, @asimshankar. Thanks for comments. I updated the code, you can have a look. I didn't fixed only concurrency things in `Server.java` because from my perspective your approach is much more complicated and error prone. Well, anyway, if you prefer it, I can use it.

Speaking about Python API, I'd be glad to propagate this change into it also, but I think it makes sense to do it in other PR (there are some our internal tasks that depends on this PR, so would be great to merge it as soon as it possible).

dmitrievanthony committed:
Update after review.

dmitrievanthony commented:
Hi, @asimshankar. I think I fixed all your comments, could you please have a look?

asimshankar commented on tensorflow/c/c_api.cc:
```diff
@@ -2803,4 +2803,40 @@ TF_Buffer* TF_GetRegisteredKernelsForOp(const char* name, TF_Status* status) {
   return ret;
 }
+
+// TF_Server functions ----------------------------------------------
```
You may want to hide all of this inside an `#ifdef __ANDROID__` similar to https://github.com/tensorflow/tensorflow/blob/1737b809b7a7b258c81a77dd4e4670315c70b53d/tensorflow/c/c_api.cc#L2249 , since we don't include the required dependencies when building for Android (as per the BUILD file changes), which is probably a good thing since there doesn't seem to be a need for this in Android builds yet.

(So, these functions should fail if invoked in Android)

dmitrievanthony replied:
Done, wrapped the whole block.

asimshankar commented on tensorflow/java/src/main/native/server_jni.cc:
```diff
@@ -0,0 +1,100 @@
+                                                        jlong handle) {
+  TF_Status* status = TF_NewStatus();
+  TF_Server* server = requireHandle(env, handle);
+  if (server == nullptr) return;
```
Create the `TF_Status` after this, otherwise we leak a `TF_Status` object.

dmitrievanthony replied:
Good point, done.

asimshankar commented on tensorflow/java/src/main/native/server_jni.cc:
```diff
@@ -0,0 +1,100 @@
+                                                       jlong handle) {
+  TF_Status* status = TF_NewStatus();
+  TF_Server* server = requireHandle(env, handle);
+  if (server == nullptr) return;
```
Create `TF_Status` after this return statement (otherwise we can leak a `TF_Status` object)

dmitrievanthony replied:
Done.

asimshankar commented on tensorflow/java/src/main/native/server_jni.cc:
```diff
@@ -0,0 +1,100 @@
+                                                       jlong handle) {
+  TF_Status* status = TF_NewStatus();
+  TF_Server* server = requireHandle(env, handle);
+  if (server == nullptr) return;
```
Same comment.

dmitrievanthony replied:
Done.

asimshankar reviewed (changes requested):
A few more leaks, otherwise looks good.

dmitrievanthony committed:
Update after review.

dmitrievanthony commented:
Thanks, I eliminated leaks. Please have a look.

dmitrievanthony committed:
Fix style.

dmitrievanthony commented:
Hi @asimshankar. Looks like tests failed:

- Android build failed because `c_api_internal.h` contained include of `distributed_runtime/server_lib.h`. I wrapped it into `#ifndef __ANDROID__`, so now it should work.
- Clang check failed on `Server.java` file, I fixed style.
- Python GPU test failed on some GPU things not related to my changes, don't think I can do something with it.
- Ubuntu Sanity failed on BUILD (I've fixed) and with the following exception:

```
Missing the licenses for the following external dependencies:
@grpc//
@grpc//third_party/address_sorting
@grpc//third_party/nanopb
```
I don't clearly understand how to fix it. Do you?

dmitrievanthony committed:
Fix style.

dmitrievanthony commented:
Hi, @ymodak, @asimshankar. I've updated the code, could you please have a look and rerun tests.

asimshankar commented:
Thanks @dmitrievanthony : I'll run the tests once we have a fix for the licenses failure.

Some background: When we package the JNI library in a tarball, we include the LICENSE files of all libraries that the target depends on. You can probably run [`ci_sanity.sh`](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/tools/ci_build/ci_sanity.sh) locally to reproduce the failure seen in the "Ubuntu Sanity" build. 

The LICENSE files are explicitly listed in the [`//tensorflow/tools/lib_package:jnilicenses_generate`](https://github.com/tensorflow/tensorflow/blob/ee1263ea3a0d8cb0bccf824d78d25d6c9d545561/tensorflow/tools/lib_package/BUILD#L181) BUILD target and the sanity check script aims to ensure that this list is in [sync with the actual dependencies of `//tensorflow/java:libtensorflow_jni.so` target](https://github.com/tensorflow/tensorflow/blob/ee1263ea3a0d8cb0bccf824d78d25d6c9d545561/tensorflow/tools/ci_build/ci_sanity.sh#L414).

Since we're adding a dependency to another third party library (grpc) we need to include the license for that as well - listed in the logs of the failing test:
```
FAIL: mismatch in packaged licenses and external dependencies
Missing the licenses for the following external dependencies:
@grpc//
@grpc//third_party/address_sorting
@grpc//third_party/nanopb
```

(Which means we need the LICENSE file for grpc as well as _its_ external dependencies on address_sorting and nanopb). 

So I think simply adding these to the dependencies of the [`//tensorflow/tools/lib_package:jnilicenses_generate`](https://github.com/tensorflow/tensorflow/blob/ee1263ea3a0d8cb0bccf824d78d25d6c9d545561/tensorflow/tools/lib_package/BUILD#L181) target should do.

Let me know if you can do that (and run `ci_sanity.sh` locally to iterate faster). If not, and you need some help, I can try to patch that in before merging this PR.

Thanks!

dmitrievanthony committed:
Fix license check.

dmitrievanthony commented:
Thanks for very details explanation, @asimshankar! I fixed the problem and checked `ci_sanity.sh` locally. Please have a look and rerun tests.

dmitrievanthony commented:
Hi @asimshankar, looks like something not related to my code failed. Some allocation problems as far as I see.

asimshankar commented:
@dmitrievanthony : Yeah, that seems unrelated. I'll take it from here :). Thanks for the contribution, will hopefully figure that out and get it merged in the next day or two. Will ping back if I discover any issues.

asimshankar commented on tensorflow/java/src/main/native/server_jni.cc:
```diff
@@ -0,0 +1,131 @@
+JNIEXPORT void JNICALL Java_org_tensorflow_Server_start(JNIEnv* env,
+                                                        jclass clazz,
+                                                        jlong handle) {
+#ifdef __ANDROID__
```
Actually, one more thing: Why not put these error messages in `c_api.cc` instead of in the JNI code?

dmitrievanthony replied:
Currently this functions are fully excluded from `c_api.cc` and `c_api.h` with their definitions in case we build TensorFlow for Android. Additionally, these exceptions are Java specific (like `NotSupportedOperationException`), so I can't throw them from `c_api.cc` anyway. Of course, I can throw another exception or status from `c_api.cc` and handle it in JNI converting into `NotSupportedOperationException`, but it looks like it will significantly complicate the solution, don't you think so?

dmitrievanthony replied:
Okay, @asimshankar, I moved these `#ifdef` constructions into `c_api.cc`, please have a look.

dmitrievanthony committed:
Update after review.

asimshankar commented on tensorflow/c/c_api.cc:
```diff
@@ -2803,4 +2803,66 @@ TF_Buffer* TF_GetRegisteredKernelsForOp(const char* name, TF_Status* status) {
+#ifdef __ANDROID__
+  status->status = tensorflow::errors::Unimplemented(
+      "Server functionality is not supported in Android");
+  return nullptr;
```
Remove this statement (since this function doesn't return anything. This breaks compilation on Android)

dmitrievanthony replied:
Yes, sorry, copy-paste is an evil :)
Fixed.

asimshankar commented on tensorflow/c/c_api.cc:
```diff
@@ -2803,4 +2803,66 @@ TF_Buffer* TF_GetRegisteredKernelsForOp(const char* name, TF_Status* status) {
+#ifdef __ANDROID__
+  status->status = tensorflow::errors::Unimplemented(
+      "Server functionality is not supported in Android");
+  return nullptr;
```
Ditto

dmitrievanthony replied:
Fixed.

asimshankar commented on tensorflow/c/c_api.cc:
```diff
@@ -2803,4 +2803,66 @@ TF_Buffer* TF_GetRegisteredKernelsForOp(const char* name, TF_Status* status) {
+#ifdef __ANDROID__
+  status->status = tensorflow::errors::Unimplemented(
+      "Server functionality is not supported in Android");
+  return nullptr;
```
Ditto

dmitrievanthony replied:
Fixed.

asimshankar commented:
(Some more comments based on the failing tests)

dmitrievanthony committed:
Update after review.

dmitrievanthony commented:
Hi, @asimshankar. Do I understand correctly that the PR will be merged soon by you or someone else? Do I need to make some additional changes?

asimshankar commented:
@dmitrievanthony : Yup, and done! Thanks for the contribution.

byronyi commented:
Hi @asimshankar, I just checked my libtensorflow v1.12 installation and see no additional C APIs introduced in this PR. @dmitrievanthony Have you been successfully building the Java part of this PR (possibly in tf-io) without source dependency on TF core?

asimshankar commented:
@byronyi : the merge commit isn't part of the 1.12 release branch since it landed after that branch was cut. So, these features will be part of 1.13 or when built from source.
